shared_model/utils/lazy.hpp

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_LAZY_HPP
+#define IROHA_LAZY_HPP
+
+#include <functional>
+#include <memory>
+
+namespace shared_model {
+  namespace detail {
+
+    /**
+     * Lazy class for lazy converting one type to another
+     * @tparam Source - input type
+     * @tparam Target - output type
+     */
+    template <typename Source, typename Target>
+    class Lazy {
```

Comment:
It's a bad name for that class. It's main purpose is converting (and actually caching) stuff, not "being lazy"
Imo it's fine to name it just `converter`

Reply:
This is not simple `converter` because converter it is `TransformType`. Maybe `LazyConverter`. @lebdron maybe you have ideas about the name?

Reply:
`LazyInitializer` sounds good to me. In case it will be reused several times in class member definition, a separate private `using` can be defined in the class: `using l = LazyInitializer<>`